### ISSUE TITLE
Check directories recursively

### ISSUE THREAD

**Issue by untitaker** [koalaman/shellcheck#143]
Thanks for this useful program! It would be nice, however, if i could just point this at my dotfiles directory and let it automatically check all shell files (recognized by shebang?)

**Comment by dialex** [comment]
For beginners, can you provide a code snippet of using `find` to achieve it?

**Comment by coudenysj** [comment]
The problem with this approach is that you cannot rely on the exit code of the find script (in a build/test script for instance).

You can use this to accomplish that:
```
for file in $(find path/to/scripts -type f); do shellcheck --format=gcc $file; done;
```

**Comment by untitaker** [comment]
The logic for detecting what is supposed to be a shellscript and what not is sometimes quite contrived. There is not a single file extension to detect files by, executable bit is not always set... I actually now agree with @koalaman, but probably for different reasons.

**Comment by coudenysj** [comment]
Oh, yes off course.

I use a recursive grep to find `#!/bin/sh` line in files, and pipe those filenames to shellcheck.

I just wanted to note the fact that if you want to use the exit codes, you need to run the commands in a loop.

**Comment by koalaman** [comment]
This topic needs a wiki page.

**Comment by fkromer** [comment]
@dialex Your snippet doesn't distinguish between `*.sh` and other files (and could lead to false positives), right?

**Comment by fkromer** [comment]
@coudenysj If I run your snippet via `gitlab-runner exec docker shellcheck`

and a `.gitlab-ci.yml`

```
shellcheck:
  image: koalaman/shellcheck-alpine
  script:
    - shellcheck --version
    - for file in $(. -type f); do shellcheck --format=gcc $file; done;
```

I get this error

```
$ for file in $(. -type f); do shellcheck --format=gcc $file; done;
/bin/sh: .: line 21: illegal option -t
```

Any idea how to fix that?

**Comment by coudenysj** [comment]
@fkromer You need to use the `find` command.

`$ for file in $(find . -type f); do shellcheck --format=gcc $file; done;`

**Comment by coudenysj** [comment]
@koalaman I've created a [Recursiveness](https://github.com/koalaman/shellcheck/wiki/Recursiveness) wiki page.

**Comment by fkromer** [comment]
@coudenysj Ooh, thanks for catching that.

**Comment by Pawamoy** [comment]
@coudenysj, about how to tell if a file is a shell script: I remembered myself using the `file` command:

```bash
for f in $(file path/to/scripts/* | grep 'shell script' | cut -d: -f1); do
  shellcheck "$f"
done
```

I think `file` uses the shebang as well to determine if a file is a shell script, and it might handle it better than a custom regex (`#\!\(/usr/bin/env \|/bin/\)sh`), though I can't confirm.

As a consequence, `file` will not detect shell scripts that do not have shebangs (like the ones we source, therefore not needing a shebang), but the `grep` search will not either. Tough, I think it's conventional to name sourced files with the `.sh` extension, or `.bash` depending on the shell being targeted.

So a combination of `find` and `file` would be:

```bash
for f in $({ find path/to/scripts -type f -regex ".*\.\w*sh"
             file path/to/scripts/* | grep 'shell script' | cut -d: -f1
           } | sort -u); do  # note that we remove potential duplicates
  shellcheck "$f"
done
```

Do you think it's worth noting in the Recursiveness page :smile: ?

**Comment by coudenysj** [comment]
Thanks for the feedback @Pawamoy (learning more every day).

This is the biggest reason `shellcheck` is not bothering with recursiveness ([Do One Thing and Do It Well](https://en.wikipedia.org/wiki/Unix_philosophy#Do_One_Thing_and_Do_It_Well)) :nerd_face:. Everyone has its own way of writing scripts. I was just trying to kickstart the Recursiveness page, I'll try to incorporate your suggestions!

A couple of questions/notes:

* @koalaman Does `shellcheck` autodetect the shell?
* If I test different files with `file` I get "mixed" results (so no "parsed" type based on shebang).
```
bin/console: a /usr/bin/env php script, ASCII text executable
bin/restart-app: a /usr/bin/env sh script, ASCII text executable
bin/test: Bourne-Again shell script, ASCII text executable
```
I will include the `file` command on the wiki page, nevertheless.

**Comment by Pawamoy** [comment]
@coudenysj You are absolutely right, finding all shell scripts is not a trivial task, and it's not in the scope of shellcheck!

About your mixed results: a better regex on `file`'s output would be `grep -E 'sh(ell)? script'`. It will catch the `/usr/bin/env` ones, but not the `Python script` ones, nor the ones that contain "sh" like Flash.

However, `file` is not perfect: it recognizes several of my scripts as C source code, even though they have a `#!/usr/bin/env bash` shebang :confounded:. It might be because of the documentation comments I use, or the `main` function, I don't know.

**Comment by coudenysj** [comment]
@Pawamoy It doesn't really matter. We don't need to cover every single possibility.

The main reason of the Recursiveness is to give people hints about how to get started.

I'll update the page with the information we have, if other use cases come out of this thread, we can update the page accordingly.

**Comment by artem-zinnatullin** [comment]
Great addition to the project wiki!

Here is a snippet i use to run it in Docker container using `shellcheck-alpine` images provided by the project:

```console
docker run \
--rm \
--volume `"pwd"`:/project:ro \
--entrypoint sh \
koalaman/shellcheck-alpine:v0.4.7 \
-c 'for file in $(find /project/ -type f -name "*.sh"); do shellcheck --format=gcc $file; done;'
```

Note that it requires Alpine image to run `sh` (regular `koalaman/shellcheck` image doesn't have `sh`) and `--entrypoint sh` to override [default one which is `shellcheck` itself](https://github.com/koalaman/shellcheck/blob/ee354ffce872a733890b356b63a96bd371c7c26f/.travis.yml#L37)

I've combined it with [wiki snippet](https://github.com/koalaman/shellcheck/wiki/Recursiveness#with-exit-code) that produces exit code, so you can use it for CI and build scripts, but of course you can customize it. 

Hope someone will find it useful! Maybe wiki should have it as well :)

**Comment by rivertam** [comment]
@artem-zinnatullin Thank you so much for this. This is exactly what I needed.

It's worth noting, however, that I think if the inner part (the command) were to actually be shellchecked, I think it would complain about using `find` output for a `for` loop ([SC2044](https://github.com/koalaman/shellcheck/wiki/SC2044)).

I modified it once more to be cleaner IMO and to fit my needs personally.

```
docker run \
  --rm \
  --volume "$(pwd)/scripts":/mnt \
  --entrypoint sh \
  koalaman/shellcheck-alpine:latest \
  -c "
exitcode=0
find /mnt -type f > /tmp/filestolint
while IFS= read -r file
do
  echo \"Checking \$file\"
  if ! shellcheck --format=gcc \"\$file\"; then
    exitcode=1
  fi
done < /tmp/filestolint
exit \$exitcode"
```

The `find` command will need to be tweaked for your needs (for example, most people use `.sh` to represent shell files, while I decided not to for this project).

**Comment by pd93** [comment]
I found @rivertam's snippet above really useful, but I wanted to use this in my CI without having to write to a temporary file, so I wrote [this script](https://gist.github.com/pd93/90e0ea11290839b5542071690c19a08c) based on it.

Note the brackets around the while loop. This keeps the `$EXITCODE` variable from going out of scope because of the pipe. It also prints the error/success in red/green respectively for readability.

**Comment by JohnVonNeumann** [comment]
First of all, I'll say that this is an annoyance, I'm not convinced on the idea of a linter without the ability to recursively run, out of the box. I get that finding bash files can be annoying, but even if you just provided default "we only search for *.sh" files, I think this would satisfy a lot of requirements.

An actually productive comment though, I used @rivertam 's script and didn't get any luck (and also felt uncomfortable with such a long command) but edited it to:

```
docker run   --rm   --volume "$(pwd)":/mnt   --entrypoint sh   koalaman/shellcheck-alpine:latest   -c "shellcheck **/*.sh"
```

And got some love.

Basically, ripped from the Travis ci docs where they bundle SC into their boxes.

Hopefully this works for other people.

Peace.

**Comment by HariSekhon** [comment]
My `check_bash_syntax.sh` script does this and a bit more by pushd'ing to each script's src directory so relative `source` directives work too, working around issues https://github.com/koalaman/shellcheck/issues/908 and https://github.com/koalaman/shellcheck/issues/1837:

https://github.com/HariSekhon/DevOps-Bash-tools

recurse all `*.sh` from `$PWD`
```
check_bash_syntax.sh
```

recurse all `*.sh` from somedir
```
check_bash_syntax.sh somedir
```

check only the files and directories I specify
```
check_bash_syntax.sh file1 file2 file3 directory1 directory2
```

The reason it's called check_bash_syntax is that I also warn for using `#!/bin/bash` instead of `#!/usr/bin/env bash`.

This script is easily modified if you wanted to auto-recurse for contents of `#!/bin/sh` for example, but I think it's a good practice to put the `.sh` extension on all scripts as it is more explicit and intuitive at a glance without needing historical project knowledge or grepping / file'ing on the contents of every file.

**Comment by corbolais** [comment]
> The problem with this approach is that you cannot rely on the exit code of the find script (in a build/test script for instance).
> 
> You can use this to accomplish that:
> 
> ```
> for file in $(find path/to/scripts -type f); do shellcheck --format=gcc $file; done;
> ```

This recommendation in a shellcheck thread is ..uh ..curious. Using the above snippet in a script yields:

    For loops over find output are fragile. Use find -exec or a while read loop. [SC2044]

cheers

**Comment by sfc-gh-cschneider** [comment]
I agree this would be easier if it was built into `shellcheck` like `shfmt` does. the main issue I ran into is that shell scripts are not always named `.sh`, so you actually have to check each script with `file` (`shfmt` does this automatically by checking for `#!/bin/sh` at the top, and it seems to ignore the `.git` dir automatically too)

Here's what seems to work for me however:
```
find . -type f -not -path "./3rd-party/*" -not -path "*/\.*" -exec file {} \+ | \
grep "shell script" | \
awk -F: '{print $1}' | \
xargs -n 20 shellcheck 
```

**Comment by cove** [comment]
I think this would be better if it was inside of `shellcheck` like `shfmt` does it by default (e.g. `shfmt -d .`). 

there's a few problems with using `find` with `shellcheck` I ran into:
* shell scripts sometimes don't have `.sh` extensions, so you have to determine if it's a shell script using `file` etc.
* if the shell script has relative paths in it, `shellcheck` needs to be run from the directory the script is in, which makes processing a git repo more complex to do efficiently

**Comment by sanmai-NL** [comment]
You can use `shfmt -f` to get a list of all shell scripts. Complex logic can be added in `shfmt`.

**Comment by jmcantrell** [comment]
This is the best I could come up with, but I think it's the best option. You may need to tailor the extension, the -s argument, and the find command:

```sh
#!/usr/bin/env bash

set -eu

while IFS= read -r -d $'\n' file; do
    if [[ $file == *.sh || $(file -b --mime-type "$file") == text/x-shellscript ]]; then
        shellcheck -W0 -s bash "$file" || continue
    fi
done < <(find . -type f \! -path "./.git/*")
```

Personally, I would put this in a `lint.sh` script in the root of your project.

**Comment by dongleivip** [comment]
> @koalaman I've created a [Recursiveness](https://github.com/koalaman/shellcheck/wiki/Recursiveness) wiki page.

Ah... I searched around and find this issue finally. However, it's been there for years... 

I'd like to suggest putting it on the Readme page.   🙂

**Comment by iarjunphp** [comment]
For Windows (PowerShell script ) -https://arjunphp.com/shellcheck-check-directories-recursively/

**Comment by brettaufheber** [comment]
./bin/entrypoint.sh:293:16: note: Double quote to prevent globbing and word splitting. [SC2086]
./bin/entrypoint.sh:294:11: note: Double quote to prevent globbing and word splitting. [SC2086]

**Comment by avidseeker** [comment]
Ironically, the [suggested script to find by shebang](https://github.com/koalaman/shellcheck/issues/143) has [SC2038](https://www.shellcheck.net/wiki/SC2038) warning.

I suppose this fixes it?

```
shellchecks() {
	find "${1:-.}" -type f -exec grep -Eq '^#!(.*/|.*env +)(sh|bash|ksh)' {} \; -exec shellcheck {} +
}
```

**Comment by navrkald** [comment]
Personally I want check all shell scripts, which are tracked in Git. For this purpose I am using [KISS](https://en.wikipedia.org/wiki/KISS_principle):

```sh
git ls-files | grep '.*.sh$' | xargs shellcheck
```

**Comment by opk12** [comment]
Proposed syntax: `-r by_extension,by_shebang`

**Comment by aep-sunlife** [comment]
Note that the lack of directory recursion in linters creates portability problems. It's not easy for users to construct bug-free recursion snippets, which would work in all shell implementations from POSIX to csh to ion to PowerShell to Command Prompt to Git Bash to MinGW to Cygwin to msysgit to fish to ...

I've got a tool to assist ShellCheck with recursion:

https://github.com/mcandre/stank

But would gladly delete it in favor of stock support for automatic directory recursion (+ following gitignores).

**Comment by septatrix** [comment]
> Note that the lack of directory recursion in linters creates portability problems. It's not easy for users to construct bug-free recursion snippets, which would work in all shell implementations from POSIX to csh to ion to PowerShell to Command Prompt to Git Bash to MinGW to Cygwin to msysgit to fish to ...
> 
> But would gladly delete it in favor of stock support for automatic directory recursion (+ following gitignores).

While native recursion support would be great, something like `find` or the more modern [`fd`](https://github.com/sharkdp/fd) are more robust than globs and the latter also respects gitignores by default

**Comment by artificial-intelligence** [comment]
> > Note that the lack of directory recursion in linters creates portability problems. It's not easy for users to construct bug-free recursion snippets, which would work in all shell implementations from POSIX to csh to ion to PowerShell to Command Prompt to Git Bash to MinGW to Cygwin to msysgit to fish to ...
> > But would gladly delete it in favor of stock support for automatic directory recursion (+ following gitignores).
> 
> While native recursion support would be great, something like `find` or the more modern [`fd`](https://github.com/sharkdp/fd) are more robust than globs and the latter also respects gitignores by default

There are a lot of environments where you can't or just straight up are not allowed to install arbitrary tools in your CI pipeline just to provide basic recursion, so the original point still stands imho.

**Comment by mcandre** [comment]
Nonrecursive scanners invite accidents.

Note that UNIX find for shell identification purposes is a bad joke. I don't think it handles shebangs well. find's notation is arcane. POSIX find has crap exit codes. For Windows native environments, it's often a nonstarter. Most of the Stack Overflow examples focus too narrowly on just "sh" at the expense of other implementations from ash to zsh. But file extension alone, basename alone, are still not sufficient to properly identify shell scripts. That's why I prototyped stank.

Frustrated that employers fail to sponsor such vital safety mechanisms as ShellCheck.

Frankly, shell scripting is inherently hazardous. Even short, innocuous scripts trigger numerous bugs. Not to mention, too many commands implement horrible exit code semantics.

I've been steadily rewriting shell glue code as dedicated Go or Rust applications. That way avoids much of the worst glitches arising from shells.

Raku has a lightweight syntax for invoking shell commands, with safe, predictable piping. There's no SCA for Raku dependencies, however. Aim for API's not CLI's.

**Comment by wileyhy** [comment]
"I've been steadily rewriting shell glue code as dedicated Go or Rust
applications. That way avoids much of the worst glitches arising from
shells."

Perhaps some things have changed? My understanding is that many fewer
people know Go & Rust as know Shell, and that both Go & Rust are
considerably more difficult to learn than Shell.

**Comment by mcandre** [comment]
Shell looks easy but is difficult to write foolproof, scaling poorly with program size.

Most non-shell languages effectively unset IFS; set -euo pipefail by default.

**Comment by wileyhy** [comment]
(You may already be aware of this information....)

Well, for what it's worth, there is the POSIX  Standard for Shell and
Utilities. They've codified existing historical practices of early shells
(Bourne, Korn, Almquist, Thompson, Tenex C, etc.) and defined a standard
that is intended to serve as a common (and therefore portable) set of
syntax for Unix `/bin/sh` and a list of ~125 to ~150 common Unix utilities.
The idea being that one can write a script on Fedora 43 in Bash 5.3 today,
and if the syntax meets the, say, POSIX-2008 standard, then there's a
reasonable, high likelihood that the script would also execute on a (still
in service) Oracle Solaris 10 box under Extended Support.

https://pubs.opengroup.org/onlinepubs/9799919799/

Cheers,
Wiley

**Comment by mcandre** [comment]
cough

**Comment by wileyhy** [comment]
And that
